mirrord/layer/src/exec.rs

```diff
@@ -96,108 +95,66 @@ fn intercept_tmp_dir(argv_arr: &Nul<*const c_char>) -> Detour<Vec<*const c_char>
-    /// functions.
-    pub struct PatchedArgs {
-        original_path: *const c_char,
-        path_c_string: Option<CString>,
-        original_argv: *const *const c_char,
-        argv_vec: Option<Vec<*const c_char>>,
-    }
-
-    impl PatchedArgs {
-        pub fn new(
-            original_path: *const c_char,
-            path_c_string: Option<CString>,
-            original_argv: *const *const c_char,
-            argv_vec: Option<Vec<*const c_char>>,
-        ) -> Self {
-            Self {
-                original_path,
-                path_c_string,
-                original_argv,
-                argv_vec,
-            }
-        }
-
-        /// Get the (maybe patched) path that should be executed.
-        pub fn get_path(&self) -> *const c_char {
-            // If we created a new string return a pointer to that, otherwise just the original
-            // pointer.
-            self.path_c_string
-                .as_ref()
-                .map(|path| path.as_ptr())
-                .unwrap_or(self.original_path)
-        }
 
-        /// Get the (maybe changed) argv.
-        pub fn get_argv(&self) -> *const *const c_char {
-            // If we created a new array return a pointer to that, otherwise just the original
-            // pointer.
-            self.argv_vec
-                .as_ref()
-                .map(|argv| argv.as_ptr())
-                .unwrap_or(self.original_argv)
-        }
+        c_string_vec.push(
+            CString::new(arg_str
+                .strip_prefix(&tmp_dir)
+                .inspect(|original_path| {
+                    trace!(
+                        "Intercepted mirrord's temp dir in argv: {}. Replacing with original path: {}.",
+                        arg_str,
+                        original_path
+                    );
+                })
+                .unwrap_or(arg_str) // No temp-dir prefix found, use arg as is.
+                // As the string slice we get here is a slice of memory allocated and managed by
+                // the user app, we copy the data and create new CStrings out of the copy without
+                // consuming the original data.
+                .to_owned()
+            )?
+        );
     }
+    Success(c_string_vec)
 }
 
 unsafe fn patch_sip_for_new_process(
     path: *const c_char,
     argv: *const *const c_char,
-) -> PatchedArgs {
+) -> (CString, Vec<CString>) {
     let exe_path = env::current_exe()
         .map(|path| path.to_string_lossy().to_string())
         .unwrap_or_default();
     trace!("Executable {} called execve/posix_spawn", exe_path);
 
-    // Do unsafe part of path conversion here.
-    let rawish_path = (!path.is_null()).then(|| CStr::from_ptr(path));
+    // TODO: should I change this function to return a result, and if exec/spawn are called with a
+    //       null pointer we could just pass it on instead of panicking?
+    let path_str = raw_to_str(&path).expect("Exec/Spawn hook got null pointer for path.");
```

Comment:
What do you think? That would mean having (duplicate) handling code in the `..._detour` functions.

How would you handle an error here?

Reply:
Hmm can't this function return a `Detour`, and on null pointer here, you return it as a `Bypass`?

Reply:
Sure. However this means we have to have (potentially duplicate) handling code in the callers of this function. Is that ok?

Reply:
Do you mean that 

```rust
let (path, argv) = patch_sip_for_new_process(path, argv);
let null_terminated_argv = c_string_vec_to_null_terminated_pointer_vec(&argv);
FN_EXECVE(path.as_ptr(), null_terminated_argv.as_ptr(), envp)
```

has to be something like

```rust
match patch_sip_for_new_process(path, argv) {
    Ok(...) => {
        let null_terminated_argv = c_string_vec_to_null_terminated_pointer_vec(&argv);
        FN_EXECVE(path.as_ptr(), null_terminated_argv.as_ptr(), envp)
    }
    Bypass(...) => {
        let null_terminated_argv = c_string_vec_to_null_terminated_pointer_vec(&argv);
        FN_EXECVE(path.as_ptr(), null_terminated_argv.as_ptr(), envp)
    }
}

```

?

Reply:
Yeah, something like that.

Reply:
If all 3 variants of `Detour` in `patch_sip_for_new_process` return the same type of value, you can do:

```rust
let Success(null_terminated_argv) | Bypass(null_terminated_argv) | Error(null_terminated_argv) = ...
FN_EXECVE(path.as_ptr(), null_terminated_argv.as_ptr(), envp)
```

Maybe you can reduce this even more, with some clever `map`-style usage, but I think this is good enough?

Reply:
I don't think they would return the same type. I mean, if they would, we wouldn't need to return a `Detour` at all, we could just return that type.

Reply:
Well, if they're all the same function call just extract it into a shared function. We kinda have this for a few detours.